nativelib/src/main/resources/gc/commix/Heap.c

```diff
@@ -0,0 +1,542 @@
+#ifdef ENABLE_GC_STATS
+    char *statsFile = Settings_StatsFileName();
+    if (statsFile != NULL) {
+        heap->stats = malloc(sizeof(Stats));
+        Stats_Init(heap->stats, statsFile, MUTATOR_THREAD_ID);
+    } else {
+#endif
+        heap->stats = NULL;
+#ifdef ENABLE_GC_STATS
+    }
+#endif
+
+
+    int gcThreadCount = Settings_GCThreadCount();
+    heap->gcThreads.count = gcThreadCount;
+    heap->gcThreads.phase = gc_idle;
+    GCThread *gcThreads = (GCThread *)malloc(sizeof(GCThread) * gcThreadCount);
+    heap->gcThreads.all = (void *)gcThreads;
+    for (int i = 0; i < gcThreadCount; i++) {
+        Stats *stats;
+#ifdef ENABLE_GC_STATS
+        if (statsFile != NULL) {
+            int len = strlen(statsFile) + 5;
+            char *threadSpecificFile = (char *) malloc(len);
+            snprintf(threadSpecificFile, len, "%s.t%d", statsFile, i);
+            stats = malloc(sizeof(Stats));
+            Stats_Init(stats, threadSpecificFile, (uint8_t)i);
+        } else {
+#endif
+            stats = NULL;
+#ifdef ENABLE_GC_STATS
+        }
+#endif
+        GCThread_Init(&gcThreads[i], i, heap, stats);
+    }
+
+    heap->mark.lastEnd_ns = scalanative_nano_time();
+
+    pthread_mutex_init(&heap->sweep.growMutex, NULL);
+}
+
+word_t *Heap_AllocLarge(Heap *heap, uint32_t size) {
+
+    assert(size % ALLOCATION_ALIGNMENT == 0);
+    assert(size >= MIN_BLOCK_SIZE);
+
+    Object *object = LargeAllocator_GetBlock(&largeAllocator, size);
+    if (object != NULL) {
+done:
+        assert(object != NULL);
+        assert(Heap_IsWordInHeap(heap, (word_t *)object));
+        return (word_t *)object;
+}
+
+    if (!Sweeper_IsSweepDone(heap)) {
+        object = Sweeper_LazySweepLarge(heap, size);
+        if (object != NULL)
+            goto done;
+    }
+
+    Heap_Collect(heap);
+
+    object = LargeAllocator_GetBlock(&largeAllocator, size);
+    if (object != NULL)
+        goto done;
+
+    if (!Sweeper_IsSweepDone(heap)) {
+        object = Sweeper_LazySweepLarge(heap, size);
+        if (object != NULL)
+            goto done;
+    }
+
+    size_t increment = MathUtils_DivAndRoundUp(size, BLOCK_TOTAL_SIZE);
+    uint32_t pow2increment = 1U << MathUtils_Log2Ceil(increment);
+    Heap_Grow(heap, pow2increment);
+
+    object = LargeAllocator_GetBlock(&largeAllocator, size);
+
+    goto done;
+}
+
+NOINLINE word_t *Heap_allocSmallSlow(Heap *heap, uint32_t size) {
+    Object *object = (Object *) Allocator_Alloc(&allocator, size);
+
+    if (object != NULL) {
+done:
+        assert(Heap_IsWordInHeap(heap, (word_t *)object));
+        assert(object != NULL);
+        ObjectMeta *objectMeta = Bytemap_Get(allocator.bytemap, (word_t *)object);
+#ifdef DEBUG_ASSERT
+        ObjectMeta_AssertIsValidAllocation(objectMeta, size);
+#endif
+        ObjectMeta_SetAllocated(objectMeta);
+        return (word_t *)object;
+    }
+
+    if (!Sweeper_IsSweepDone(heap)) {
+        object = Sweeper_LazySweep(heap, size);
```

Comment:
`Sweeper_LazySweep` needs a better name. Based on the code it's more like `Sweeper_LazySweepUntilCanAllocate`.

Reply:
Sweeping until it is possible to allocate is literally the definition of lazy sweeping. `UntilCanAllocate` is redundant.